### PR TITLE
nimony: implements `genEnumToStrProc`

### PR DESCRIPTION
**Pull request by ringabout**
ref https://github.com/nim-lang/nimony/issues/288
```
 (proc :dollar.Color1.0.tesg7afhq1 x . . 
  (params 
   (param :e . . Color1.0.tesg7afhq1 .))
  (string) . . 
  (stmts 9
   (result :result.0 . .
    (string) .) 9
   (case e ~7,1
    (of 
     (set Red1.0.tesg7afhq1) 
     (stmts 
      (ret "Red1"))) ~1,1
    (of 
     (set Blue1.0.tesg7afhq1) 
     (stmts 
      (ret "Blue1"))) 6,1
    (of 
     (set Green1.0.tesg7afhq1) 
     (stmts 
      (ret "Green1"))))
   (ret result.0)))
```

### CODE REVIEW

**Comment by Araq** [comment]
This should be part of Nimony. To be generated for every enum that we encounter.

File: src/nimony/sem.nim [hunk 1]
```diff
@@ -1723,12 +1723,13 @@ type
...
 proc semEnumField(c: var SemContext; n: var Cursor; state: var EnumTypeState)
 
-proc semEnumType(c: var SemContext; n: var Cursor; enumType: SymId; beforeExportMarker: int) =
+proc semEnumType(c: var SemContext; n: var Cursor; enumType: SymId; beforeExportMarker: int): bool =
   # XXX Propagate hasHole somehow
   takeToken c, n
   let magicToken = c.dest[beforeExportMarker]
   var state = EnumTypeState(enumType: enumType, thisValue: createXint(0'i64), hasHole: false,
     isBoolType: magicToken.kind == ParLe and pool.tags[magicToken.tagId] == $BoolT)
+  result = not state.isBoolType
```
Comment (Araq) [inline src/nimony/sem.nim:1732]: there is nothing wrong with doing the same for `bool`...

File: src/nimony/enumtostr.nim [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    dest.add tagToken("ret", enumDeclInfo)
+    dest.add strToken(fieldValue, enumDeclInfo)
+    dest.addParRi() # ret
+    dest.addParRi() # stmts
+
+    dest.addParRi() # of
+
+  dest.addParRi() # case
+
+proc genEnumToStrProc*(dest: var TokenBuf, typeDecl: var Cursor; stringType: var Cursor) =
```
Comment (Araq) [inline src/nimony/enumtostr.nim:48]:
```suggestion
proc genEnumToStrProc*(dest: var TokenBuf, typeDecl: var Cursor; stringType: Cursor) =
```

File: src/nimony/sem.nim (hunk 1 above)
Comment (ringabout) [reply src/nimony/sem.nim]: I ran into weird problems with `bool` with `declared` last time, the file name didn't seem to exist. Probably it gets fixed by introducing `system.nim`

File: src/nimony/sem.nim [hunk 3]
```diff
@@ -3001,8 +3004,20 @@ proc semTypeSection(c: var SemContext; n: var Cursor) =
...
   publish c, delayed.s.name, declStart
 
+  if isEnumTypeDecl:
+    var dest = createTokenBuf()
+    var enumTypeDecl = cursorAt(c.dest, declStart)
+    genEnumToStrProc(dest, enumTypeDecl, c.types.stringType)
+    endRead(c.dest)
+    var dollorProcDecl = beginRead(dest)
+    var it = Item(n: dollorProcDecl, typ: c.types.autoType)
+    semExpr(c, it)
```
Comment (Araq) [inline src/nimony/sem.nim:3018]: A bit ugly as the thing is so simple that it shouldn't require semchecking. But alright for now.

**Pull request merged**